Updated examples

UebelAndre opened the pull request:
Following up on #190 

Just in case there isn't more to it than running `smoke-test.sh`, I did that and these are the results

The only additional change was adding `shallow_since` to `smoke-test.sh` for `rules_rust`

samschlegel commented on examples/WORKSPACE:
```diff
@@ -1,24 +1,21 @@
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
 
-http_archive(
+git_repository(
```
This should maybe be changed to `http_archive` in the smoke test, as it's generally recommended to use `http_archive` over `git_repository` where possible, since it tends to be smaller, faster, and can utilize the repository cache.

UebelAndre replied:
Updated

samschlegel commented on examples/WORKSPACE:
```diff
@@ -1,24 +1,21 @@
-        # Master branch as of 2020-06-06
-        "https://github.com/bazelbuild/rules_rust/archive/6e5fa2c570ac2f17ac1df840d060fc8aab521a07.tar.gz",
-    ],
+    commit = "fda9a1ce6482973adfda022cadbfa6b300e269c3",
```
I'd also add a comment that this is the master branch as of 2020-06-30

UebelAndre replied:
Added

samschlegel commented on examples/WORKSPACE:
```diff
@@ -1,24 +1,21 @@
-    url = "https://github.com/bazelbuild/bazel-skylib/archive/1.0.0.tar.gz",
+    sha256 = "eb5c57e4c12e68c0c20bc774bfbc60a568e800d025557bc4ea022c6479acc867",
+    strip_prefix = "bazel-skylib-0.6.0",
+    url = "https://github.com/bazelbuild/bazel-skylib/archive/0.6.0.tar.gz",
```
Should this just be updated in the smoke-test? It was bumped in a PR a while back for the examples, and it probably should have been bumped in the smoke-test then too.

UebelAndre replied:
I updated `smoke-test.sh` so now it should be using the updated version of `bazel_skylib` for tests and match the example output. Sorry about not catching this before. I blindly ran `smoke-test.sh` and just committed whatever was generated 😅

UebelAndre commented:
@samschlegel Hey, I thought I should update the `README` for this update as well. Let me know what you think 😄

UebelAndre commented:
@samschlegel Hey, can I get your eyes on this real quick? I'm hoping to get this and #199 merged since I think they're pretty big quality of life improvements for this project. And before I start pining people I'd like to have a completed review 😅

samschlegel commented on smoke-test.sh:
```diff
@@ -26,41 +26,59 @@ cp -r "$TEST_DIR/remote" "$TEST_DIR/vendored" "$EXAMPLES_DIR"
 done
 
+# Define a Bazel version file
+cat > "$EXAMPLES_DIR/.bazelversion" << EOF
```
This isn't used by CI, so I have a feeling this will mostly cause confusion if it isn't kept up to date with the version set here https://github.com/google/cargo-raze/blob/eac33e62a5895fdd73630ae2edd2b456b2f72595/.travis.yml#L17

If this change is gonna be made, it might make sense to have `.travis.yml` do something like `V="$(cat examples/.bazelversion)"` instead

UebelAndre replied:
Good call. I updated this but also just added a `.bazelversion` file to the root of the repo to be the single source of truth. This should hopefully be clearer for everyone.

UebelAndre commented on smoke-test.sh:
```diff
@@ -90,6 +108,8 @@ done
 # Run the Bazel build for all targets
 cd "$EXAMPLES_DIR"
+echo "Cleaning Bazel workspace"
+bazel clean --expunge
```
Yeah, `bazel clean --expunge` would definitely slow down test iteration speed. It shouldn't matter in CI but thought it was more correct to have. I can remove it though

UebelAndre replied:
I removed this. Builds should effectively be clean builds in CI. Good call.

samschlegel replied:
`bazel clean` shouldn't ever be required for correctness and if it is, it's considered a bug by the Bazel team, so I think not relying on it seems ok

UebelAndre commented on README.md:
```diff
@@ -136,28 +168,33 @@ This tells Raze not to expect the dependencies to be vendored and to generate
 using the files generated into //cargo.
 
-*Note that this method's name depends on your `gen_workspace_prefix` setting*.
+_Note that this method's name depends on your `gen_workspace_prefix` setting_.
```
There is no visual difference here. `_` and `*` are interchangeable so this was indeed something that my editor did. I can revert it if you feel strongly

samschlegel replied:
Ah yeah, I mostly commented on this due to the change it made on line 272

UebelAndre replied:
Yeah, that one was a bad change. Thanks for catching it 😄

UebelAndre commented on examples/WORKSPACE.bazel:
```diff
@@ -0,0 +1,47 @@
+workspace(name = "cargo_raze_examples")
```
Why does it make it hard to review? This change shouldn't have any noticeable impact for the most part. I was trying to make it more unique/identifiable. But in looking at [the docs](https://docs.bazel.build/versions/master/skylark/lib/globals.html#workspace), it should maybe be updated to `com_github_google_cargo_raze_examples` or something(?). Any suggestions?

samschlegel replied:
I meant the file, not this line specifically, sorry. No way to add a comment on the whole file 😓 

I meant it's harder to review since Github just sees the old one as deleted and the new one as created.

UebelAndre replied:
OOOOOH, because I renamed it to `WORKSPACE.bazel`! I see now! 

Below is a comparison between master and this branch. Does this help or do you still feel that I should rename the file back to `WORKSPACE`?
```diff
diff --git a/examples/WORKSPACE b/examples/WORKSPACE
index 11bac71..1005541 100644
--- a/examples/WORKSPACE
+++ b/examples/WORKSPACE
@@ -1,33 +1,47 @@
-workspace(name = "examples")
+workspace(name = "com_github_google_cargo_raze_examples")
 
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
-    name = "io_bazel_rules_rust",
-    sha256 = "a62597502c1ee14fcc30da92fc31c69bc6b436f464939449dc7be88c76218a31",
-    strip_prefix = "rules_rust-6e5fa2c570ac2f17ac1df840d060fc8aab521a07",
+    name = "bazel_skylib",
+    sha256 = "97e70364e9249702246c0e9444bccdc4b847bed1eb03c5a3ece4f83dfe6abc44",
     urls = [
-        # Master branch as of 2020-06-06
-        "https://github.com/bazelbuild/rules_rust/archive/6e5fa2c570ac2f17ac1df840d060fc8aab521a07.tar.gz",
+        "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.0.2/bazel-skylib-1.0.2.tar.gz",
+        "https://github.com/bazelbuild/bazel-skylib/releases/download/1.0.2/bazel-skylib-1.0.2.tar.gz",
     ],
 )
 
+load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
+
+bazel_skylib_workspace()
+
 http_archive(
-    name = "bazel_skylib",
-    sha256 = "9a737999532daca978a158f94e77e9af6a6a169709c0cee274f0a4c3359519bd",
-    strip_prefix = "bazel-skylib-1.0.0",
-    url = "https://github.com/bazelbuild/bazel-skylib/archive/1.0.0.tar.gz",
+    name = "io_bazel_rules_rust",
+    sha256 = "5ed804fcd10a506a5b8e9e59bc6b3b7f43bc30c87ce4670e6f78df43604894fd",
+    strip_prefix = "rules_rust-fdf9655ba95616e0314b4e0ebab40bb0c5fe005c",
+    urls = [
+        # Master branch as of 2020-07-27
+        "https://github.com/bazelbuild/rules_rust/archive/fdf9655ba95616e0314b4e0ebab40bb0c5fe005c.tar.gz",
+    ],
 )
 
-load("@io_bazel_rules_rust//:workspace.bzl", "bazel_version")
-bazel_version(name = "bazel_version")
-
 load("@io_bazel_rules_rust//rust:repositories.bzl", "rust_repositories")
+
 rust_repositories()
 
+load("@io_bazel_rules_rust//:workspace.bzl", "bazel_version")
+
+bazel_version(name = "bazel_version")
+
 load("//remote/non_cratesio/cargo:crates.bzl", "non_cratesio_fetch_remote_crates")
+
 non_cratesio_fetch_remote_crates()
 
 load("//remote/complicated_cargo_library/cargo:crates.bzl", "complicated_cargo_library_fetch_remote_crates")
+
 complicated_cargo_library_fetch_remote_crates()
 
+load("//remote/regression_test/cargo:crates.bzl", "regression_test_fetch_remote_crates")
+
+regression_test_fetch_remote_crates()
+
```

samschlegel replied:
I think it's probably worth keeping as WORKSPACE, since I don't think it provides a ton of value to rename, and it will probably break the git history

UebelAndre replied:
Ok, I've restored the filename

Rant: Man, I hate `WORKSPACE` and `BUILD` so much. I really wish Bazel only supported the filenames with extensions. But oh well 😜

UebelAndre commented on README.md:
````diff
@@ -269,7 +311,7 @@ extra_aliased_targets = [
 ]
 ```
 
-Cargo-raze prefixes binary targets with "cargo_bin_", as although Cargo permits
````
Ok, this was definitely my editor being dumb 😜

UebelAndre commented:
@samschlegel Ok, I think I've addressed all the feedback minus the open discussions. Let me know if there's any more and if I've satisfied your concerns there.

samschlegel commented on examples/.bazelversion:
```diff
@@ -0,0 +1 @@
+3.4.1
```
Is this needed if it's at the root now?

UebelAndre replied:
It's a "nice to have". It means you can take the examples directory and more or less treat it as it's own workspace. I only added `.bazelversion` to the root of the repo because that's where I'd expect it to be. It just gets copied there from the file in the repo root. I can delete it if you feel it shouldn't exist. (I'm torn so I'll go with whatever makes everyone happy 😄 )

samschlegel replied:
I don't have a strong feeling either way, but my worry is that it's a bit confusing. Nothing outside of examples uses `.bazelversion`, and it feels like having two implies that the repo-wide version might differ from the examples version.

UebelAndre replied:
Ok, I removed it from the examples but left the top level `.bazelversion` file remains.

UebelAndre commented:
@samschlegel I addressed some feedback but also squashed the changes. I'm not sure how GitHub pull requests handles merges so just wanted to be safe. Sorry if that just makes it difficult to review 😅

UebelAndre commented:
@samschlegel Ok, I think everything should be good now 🙏

UebelAndre commented:
@dfreese @damienmg Would anyone with merging powers also be able to look at this and potentially merge it? 🙏

damienmg commented:
Thanks for putting that up :)

UebelAndre commented:
Yay! Thanks @damienmg! I think the next thing to look at would be #199

damienmg commented:
I'll keep in mind but I'd like to investigate a couple more bugs that I
left unanswered for so long next.
